### PR TITLE
Replace std::auto_ptr with std::unique_ptr

### PR DESCRIPTION
**Pull request by bwalden**
As part of this, some classes which previously had a destructive copy constructor and assignment operator have had these explicitly deleted and replaced with a move constructor and assignment operator. Namely `resip::Helper::ContentsSecAttrs` and `dum::TargetCommand`.

The `dum::TargetCommand` class still has a destructive `clone()` method. This doesn't appear to cause a problem, the way it is used. But it may not be semantically correct.

I have not changed anything within the contrib directory or the p2p or tfm sub projects. These can be updated if required, but are not built by either the Travis script for Linux or the Visual Studio solution for Windows.

### CODE REVIEW

File: apps/ichat-gw/MediaRelay.cxx [hunk 1]
```diff
@@ -330,10 +332,10 @@ MediaRelay::checkKeepalives(MediaRelayPort* relayPort)
       //}
 
       // Add message to buffer
-      std::auto_ptr<char> buffer(new char[sizeof(RtpHeader)]); 
+      std::unique_ptr<char> buffer(new char[sizeof(RtpHeader)]);
```
Comment (gjasny) [inline apps/ichat-gw/MediaRelay.cxx:335]: do you have to use `std::unique_ptr<char[]>` here to invoke `delete []`?

File: resip/dum/TargetCommand.cxx [hunk 2]
```diff
@@ -2,31 +2,27 @@
...
 void TargetCommand::executeCommand()
 {
-   mTarget.post(mMessage);
+   mTarget.post(std::move(mMessage));
 }
 
 Message* TargetCommand::clone() const
 {
-   return new TargetCommand(*this);
+   return new TargetCommand(mTarget, std::move(mMessage)); // !bw! this destructive clone is actually a move, which may not be right
```
Comment (gjasny) [inline resip/dum/TargetCommand.cxx:25]: Is it actually used?

File: resip/stack/Helper.cxx [hunk 3]
```diff
@@ -2206,12 +2182,12 @@ auto_ptr<SdpContents> Helper::getSdp(Contents* tree)
...
       {
          DebugLog(<< "Got sdp" << endl);
-         return auto_ptr<SdpContents>(static_cast<SdpContents*>(sdp->clone()));
+         return unique_ptr<SdpContents>(static_cast<SdpContents*>(sdp->clone()));
       }
    }
 
    //DebugLog(<< "No sdp" << endl);
-   return emptysdp;
+   return unique_ptr<SdpContents>();
```
Comment (gjasny) [inline resip/stack/Helper.cxx:2190]: `return nullptr`?

**Comment by gjasny** [comment]
The autoconf should check for C++11 compatibility and enable it. Maybe @dpocock could help there?

File: resip/stack/Helper.cxx (hunk 3 above)
Comment (bwalden) [reply resip/stack/Helper.cxx:2190]: Agreed, nullptr would look tidier.

File: resip/dum/TargetCommand.cxx (hunk 2 above)
Comment (bwalden) [reply resip/dum/TargetCommand.cxx:25]: I think so. The `dum::EventDispatcher` and `post()` methods of `SipStack` both clone messages.

File: apps/ichat-gw/MediaRelay.cxx (hunk 1 above)
Comment (bwalden) [reply apps/ichat-gw/MediaRelay.cxx:335]: Yes, I believe so.

**Comment by bwalden** [comment]
> The autoconf should check for C++11 compatibility and enable it. Maybe @dpocock could help there?

I assume we will have to add the following somewhere in configure.ac.
https://www.gnu.org/software/autoconf-archive/ax_cxx_compile_stdcxx.html

**Comment by armatusmiles** [comment]
Hello. Any updates?

**Comment by bwalden** [comment]
I think I've now made all of the suggested changes.

Possibly the only comment outstanding from issue #147, would be to also replace resip::SharedPtr with std::shared_ptr at the same time. This is certainly sensible and something I can do as part of this PR if it is wanted.

**Comment by sgodin** [comment]
Thanks Ben!

It seems that no one is opposed to changing resip::SharedPtr to
std::shared_ptr, so I think it would be great to get this in the same PR,
since both changes will likely require app updates.

Scott

Scott

**Comment by omnisip** [comment]
I've been wanting to get rid of resip::SharedPtr for years.  Please make
the change!  :-)

**Comment by bwalden** [comment]
OK so resip::SharedPtr is now replaced with std::shared_ptr. I also took the liberty of replacing boost::shared_ptr with std::shared_ptr while I was in there.

Again, I've not touched the p2p or tfm sub projects. Does anyone have any thoughts on whether these should be updated?

**Comment by portsip** [comment]
@bwalden 
Hi Bwalden, thanks for your works. Today I got the error when I try to compile your PR148 with Debian 10.4(64bit), GCC 8.3, please see this link: https://github.com/resiprocate/resiprocate/issues/147#issuecomment-677610287

BR

**Comment by sgodin** [comment]
Hi Ben,

Thank you so much for all this work.  I was hoping to connect with you via email.  Can you please drop me a message at: s g o d i n @ s i p s p e c t r u m . c o m

Scott

**Comment by bwalden** [comment]
Thank you all. It's nearly there now. I just want to finish off p2p and tfm. I decided to update these as well after finding they no longer compile following my changes.

**Comment by portsip** [comment]
@sgodin @bwalden  Hi, a question, in the ClientPublish, since the req is a shared_ptr, why use std::move on it? If use std::move there then I suggest using std::unique_ptr. If use std::shared_ptr as the origin implementation (I mean before c+11), the std::move is not necessary.

**Comment by sgodin** [comment]
> @sgodin @bwalden Hi, a question, in the ClientPublish, since the req is a shared_ptr, why use std::move on it? If use std::move there then I suggest using std::unique_ptr. If use std::shared_ptr as the origin implementation (I mean before c+11), the std::move does not make sense.

@portsip @bwalden  PortSIP - I agree with you.  Ben do you know why you used std::move here?  Did you do this elsewhere with shared_ptr's?  It could have unintended side effects.

**Comment by bwalden** [comment]
You h> > @sgodin @bwalden Hi, a question, in the ClientPublish, since the req is a shared_ptr, why use std::move on it? If use std::move there then I suggest using std::unique_ptr. If use std::shared_ptr as the origin implementation (I mean before c+11), the std::move does not make sense.
> 
> @portsip @bwalden PortSIP - I agree with you. Ben do you know why you used std::move here? Did you do this elsewhere with shared_ptr's? It could have unintended side effects.

Yes this is using the modern C++ pass-by-value pattern. Any object which is passed by value to a constructor and is to be persisted as a member variable can be moved. This allows the compiler to optimize for us. If the object is an lvalue, the object is only copied once instead of twice. If the object is an rvalue, zero copies take place.

Some info on this in this example:
https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html#:~:text=Pass%2Dby%2Dvalue%20in%20constructors,necessary%20to%20add%20this%20include.

The example uses std::string, but the principle for std::shared_ptr is the same.

```cpp
class MyFirstType
{
public:
    explicit MyFirstType(std::shared_ptr<Foo> foo)
        : m_foo(std::move(foo))
       // From this point on, foo is invalid and m_foo should be used
    {
    }

private:
    std::shared_ptr<Foo> m_foo;
};

void MyFirstFunc()
{
    auto foo1 = std::make_shared<Foo>();
    // foo1 is valid and has a ref count of 1

    MyFirstType mft1(foo1);
    // foo1 is still value and now has a ref count of 2

    auto foo2 = std::make_shared<Foo>();
    // foo2 is valid and has a ref count of 1

    MyFirstType mft2(std::move(foo2));
    // foo2 is invalid, mft2's m_foo member is has the only reference
}
```

Remember that performing a std::move on a std::shared_ptr is applying move semantics to the std::shared_ptr object itself, not the value pointed to by the std::shared_ptr.

**Comment by sgodin** [comment]
Thank you Ben, for all your work with this pull request - it has now been merged with master.

**Comment by dpocock** [comment]
@bwalden @sgodin there are some issues with the use of std::move here

I fixed one in recon and then I did a grep on reflow and found another one:

https://github.com/resiprocate/resiprocate/commit/7bb07cee4ac257dd7477adfa44836414da2bf285#diff-b3b2761215e678b972bd1b483df12aadccdff55fcc0db530bf7eb1bc60a4c2aa

In the case in recon: after the ```mProfile(std::move(profile))```, we can't use ```profile``` again, it is no longer valid, we can only use ```mProfile``` after that.  So I changed ```profile``` to ```mProfile``` in each case.

In the case in reflow: ```context``` is used twice in that branch of the ```if``` statement.  We can't use ```std::move``` twice or the second attempt is not valid.

On my platform the mProfile issue caused a segmentation fault

I'm concerned that there may be other examples of this lurking in the code but I didn't have time to do a full analysis.

**Comment by dpocock** [comment]
Added this in Bugzilla as well so I can track it for the next release

https://www.resiprocate.org/bugzilla/show_bug.cgi?id=172

**Comment by sgodin** [comment]
Yeah I was worried about the same thing.  In the few cases I found it was
obvious when running the unit tests as they caused exceptions when
misused.  I did a quick look around the code at the time and ran all our
unit tests and tfm tests.  I guess I missed a few cases.  We don't really
have any automated tests for recon, so that's probably how those went
unnoticed.  I've used the core stack and dum from master in a few projects
and haven't had issues, but I have done much with return, reflow or recon
since this work was completed.

Scott

**Comment by dpocock** [comment]
Yes, I can see it was a lot of work and everybody who contributed to this deserves a lot of credit for helping modernize the project.

I just made an unrelated fix in recon and there was a non-deterministic build failure in rutil, only on gcc, the clang compile was fine:

https://travis-ci.org/github/resiprocate/resiprocate/builds/767161345

**Comment by bwalden** [comment]
@dpocock @sgodin cast my eye over this again and I think I've found at least one other problem. Will fix and create a pull request.

**Pull request merged**